repo/blob/gdrive/gdrive_storage.go

```diff
@@ -0,0 +1,460 @@
+
+	if length != 0 {
+		// nolint:wrapcheck
+		if err := iocopy.JustCopy(output, res.Body); err != nil {
+			return translateError(err)
+		}
+	}
+
+	// nolint:wrapcheck
+	return blob.EnsureLengthExactly(output.Length(), length)
+}
+
+func (gdrive *gdriveStorage) GetMetadata(ctx context.Context, b blob.ID) (blob.Metadata, error) {
+	fileId, ok := gdrive.fileIdCache[b]
+	var file *drive.File
+	var err error
+
+	if ok {
+		file, err = gdrive.client.Files.Get(fileId).Fields(getMetadataFields).Context(ctx).Do()
+		err = translateError(err)
+	} else {
+		file, err = gdrive.QueryForBlob(ctx, b, listMetadataFields)
+	}
+	if err != nil {
+		return blob.Metadata{}, errors.Wrap(err, "get blob metadata")
+	}
+
+	bm, err := parseBlobMetadata(file, b)
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	return bm, nil
+}
+
+func (gdrive *gdriveStorage) PutBlob(ctx context.Context, b blob.ID, data blob.Bytes, opts blob.PutOptions) error {
+	switch {
+	case opts.HasRetentionOptions():
+		return errors.Wrap(blob.ErrUnsupportedPutBlobOption, "blob-retention")
+	}
+
+	fileId, err := gdrive.GetFileId(ctx, b)
+	existingFile := true
+	if errors.Is(err, blob.ErrBlobNotFound) {
+		existingFile = false
+	} else if err != nil {
+		return err
+	}
+	if existingFile && opts.DoNotRecreate {
+		return blob.ErrBlobAlreadyExists
+	}
+
+	var file *drive.File
+	mtime := ""
+
+	if !opts.SetModTime.IsZero() {
+		mtime = opts.SetModTime.Format(time.RFC3339)
+	}
+
+	if !existingFile {
+		file, err = gdrive.client.Files.Create(&drive.File{
+			Name:         toFileName(b),
+			Parents:      []string{gdrive.folderId},
+			MimeType:     gdriveMimeType,
+			ModifiedTime: mtime,
+		}).Media(data.Reader(),
+			googleapi.ChunkSize(uploadChunkSize),
+			googleapi.ContentType(uploadContentType),
+		).Context(ctx).Do()
+	} else {
+		file, err = gdrive.client.Files.Update(fileId, &drive.File{
+			ModifiedTime: mtime,
+		}).Media(
+			data.Reader(),
+			googleapi.ChunkSize(uploadChunkSize),
+			googleapi.ContentType(uploadContentType),
+		).Context(ctx).Do()
+	}
+	if err != nil {
+		return translateError(err)
+	}
+
+	gdrive.fileIdCache[b] = file.Id
+	if opts.GetModTime != nil {
+		if mtime, err := parseModifiedTime(file); err == nil {
+			*opts.GetModTime = mtime
+		}
+	}
+
+	return nil
+}
+
+func (gdrive *gdriveStorage) DeleteBlob(ctx context.Context, b blob.ID) error {
+	fileId, err := gdrive.GetFileId(ctx, b)
+	if errors.Is(err, blob.ErrBlobNotFound) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	err = gdrive.client.Files.Delete(fileId).Context(ctx).Do()
+	return translateError(err)
+}
+
+func (gdrive *gdriveStorage) ListBlobs(ctx context.Context, prefix blob.ID, callback func(blob.Metadata) error) error {
+	// Tracks blob matches in cache but not returned by API.
+	unvisitedIds := make(map[blob.ID]bool)
+
+	consumer := func(files *drive.FileList) error {
+		for _, file := range files.Files {
+			blobId := toBlobId(file.Name)
+			gdrive.fileIdCache[blobId] = file.Id
```

Comment:
cache must be thread/goroutine safe, so you'd need a mutex inside cache implementation - simple map access will generally panic on concurrent access attempts.

Reply:
Done.